File: core/src/main/java/com/timepath/vfs/provider/zip/ZipFileProvider.java
package com.timepath.vfs.provider.zip;

import com.timepath.vfs.MockFile;
import com.timepath.vfs.SimpleVFile;
import com.timepath.vfs.VFSStub;
import com.timepath.vfs.VFile;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.io.*;
import java.util.logging.Level;
import java.util.logging.Logger;
import java.util.regex.Pattern;
import java.util.zip.ZipEntry;
import java.util.zip.ZipInputStream;

public class ZipFileProvider extends VFSStub {

    private static final Logger LOG = Logger.getLogger(ZipFileProvider.class.getName());
    private static final Pattern DIRECTORYSPLIT = Pattern.compile(VFile.SEPARATOR);

    public ZipFileProvider(@NotNull byte[] data) throws IOException {
        @NotNull BufferedInputStream bis = new BufferedInputStream(new ByteArrayInputStream(data));
        bis.mark(data.length);
        @NotNull ZipInputStream zs = new ZipInputStream(bis);
        @NotNull byte[] buffer = new byte[2048];
        for (ZipEntry e; (e = zs.getNextEntry()) != null; ) {
            LOG.log(Level.FINE, "{0}", e);
            @NotNull ByteArrayOutputStream baos = new ByteArrayOutputStream((int) Math.max(e.getCompressedSize(), 0));
            for (int len; (len = zs.read(buffer)) > 0; ) {
                baos.write(buffer, 0, len);
            }
            String[] split = DIRECTORYSPLIT.split(e.getName());
            @Nullable SimpleVFile dir = this;
            // create transient directories
            for (int i = 0; i < (split.length - 1); i++) {
                String dirName = split[i];
                @Nullable SimpleVFile sub = dir.get(dirName);
                if (sub == null) {
                    sub = new MockFile(dirName);
                    dir.add(sub);
                }
                dir = sub;
            }
            dir.add(new ZipFile(e, baos.toByteArray()));
        }
    }

}

File: core/src/main/kotlin/com/timepath/vfs/provider/zip/ZipFileProvider.kt
package com.timepath.vfs.provider.zip

import com.timepath.vfs.MockFile
import com.timepath.vfs.SimpleVFile
import com.timepath.vfs.provider.ProviderStub

import java.io.BufferedInputStream
import java.io.ByteArrayInputStream
import java.io.ByteArrayOutputStream
import java.io.IOException
import java.util.logging.Level
import java.util.logging.Logger
import java.util.zip.ZipInputStream
import com.timepath.vfs.VFile

public class ZipFileProvider [throws(javaClass<IOException>())](data: ByteArray) : ProviderStub() {

    {
        val bis = BufferedInputStream(ByteArrayInputStream(data))
        bis.mark(data.size())
        ZipInputStream(bis).use { zs ->
            val buffer = ByteArray(BUFFER_SIZE)
            while (true) {
                val e = zs.getNextEntry()
                if (e == null) break
                LOG.log(Level.FINE, "{0}", e)
                var sizeLong = e.getSize()
                if (sizeLong == -1L) sizeLong = e.getCompressedSize()
                if (sizeLong == -1L) sizeLong = BUFFER_SIZE.toLong()
                if (sizeLong > Integer.MAX_VALUE) {
                    LOG.log(Level.SEVERE, "ZipEntry exceeds sizeof int")
                    continue
                }
                ByteArrayOutputStream(sizeLong.toInt()).use { baos ->
                    while (true) {
                        val len = zs.read(buffer)
                        if (len < 0) break
                        baos.write(buffer, 0, len)
                    }
                    atPath(e.getName()).add(ZipFile(e, baos.toByteArray()))
                }
            }
        }
    }

    private fun atPath(path: CharSequence): SimpleVFile {
        val split = VFile.SEPARATOR_PATTERN.split(path)
        var dir: SimpleVFile = this
        for (i in (split.size() - 1).indices) {
            val dirName = split[i]
            var sub = dir[dirName]
            // Create transient directories
            if (sub == null) {
                sub = MockFile(dirName)
                dir.add(sub!!)
            }
            dir = sub!!
        }
        return dir
    }

    class object {

        private val LOG = Logger.getLogger(javaClass<ZipFileProvider>().getName())
        private val BUFFER_SIZE = 2048
    }

}
